File: CPP/ByTier/C/1895_C_Torn_Lucky_Ticket.cpp
// Problem: CF 1895 C - Torn Lucky Ticket
// https://codeforces.com/contest/1895/problem/C

/*
Purpose: This code solves the problem of counting pairs of ticket strings such that their concatenation forms a "lucky ticket".
A lucky ticket has an even length and equal sums in the first and second halves.
The approach uses pre-processing and a hash-like counting technique to efficiently compute the answer.

Algorithms/Techniques:
- Preprocessing: Store counts of prefix sums for each possible string length and total sum
- Enumeration: Iterate over all possible pairings of lengths and compute valid combinations
- Optimization: Use a 4D array to store counts of (length, total_sum, prefix_length, prefix_sum)

Time Complexity: O(n * L^2 * S^2) where L is max length (5) and S is max sum (5 * 9 = 45)
Space Complexity: O(L^2 * S^2) for the 4D array storage
*/

#include <time.h>
#include <iostream>
#include <string>

using namespace std;
#define fio                                                                    \
  ios_base::sync_with_stdio(false);                                            \
  cin.tie(NULL);
using i64 = long long;
const i64 inf = 0x3fffffffffffffff;
i64 book[6][60][6][60]; // 4D array to store counts of (length, total_sum, prefix_len, prefix_sum)
int main() {
  fio;
  i64 n;
  cin >> n;
  string s;
  for (int i = 1; i <= n; i++) {
    cin >> s;
    i64 sum = 0;
    i64 sz = s.size();
    for (i64 j = 0; j < sz; j++) {
      sum += (s[j] - '0'); // Calculate total sum of digits
    }
    i64 prefix_sum = 0;
    for (i64 j = 0; j < sz; j++) {
      prefix_sum += (s[j] - '0'); // Compute prefix sum
      book[sz][sum][j + 1][prefix_sum]++; // Increment count in 4D array
    }
  }
  i64 ans = 0;
  for (i64 l = 1; l <= 5; l++) {
    for (i64 r = 1; r <= 5; r++) {
      if ((l + r) % 2 != 0) {
        continue; // Skip odd total lengths
      }
      if (l < r) {
        // Case where first string is shorter than second
        for (i64 lval = 1; lval <= 50; lval++) {
          for (i64 rval = 1; rval <= 50; rval++) {
            if ((lval + rval) % 2 != 0) {
              continue; // For valid half sums, both must be even or both odd
            }
            ans += book[l][lval][l][lval] *
                   book[r][rval][(r - l) / 2][(rval - lval) / 2];
          }
        }
      } else if (l == r) {
        // Case where both strings have same length
        for (i64 val = 1; val <= 50; val++) {
          ans += book[l][val][l][val] * book[l][val][l][val];
        }
      } else if (l > r) {
        // Case where first string is longer than second
        for (i64 lval = 1; lval <= 50; lval++) {
          for (i64 rval = 1; rval <= 50; rval++) {
            if ((lval + rval) % 2 != 0) {
              continue;
            }
            ans += book[l][lval][l - (l - r) / 2][lval - (lval - rval) / 2] *
                   book[r][rval][r][rval];
          }
        }
      }
    }
  }
  cout << ans << endl;
  cerr << "Time : " << (double)clock() / (double)CLOCKS_PER_SEC << "s\n";
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/C/1895_C_Torn_Lucky_Ticket.py
# Problem: CF 1895 C - Torn Lucky Ticket
# https://codeforces.com/contest/1895/problem/C

"""
Algorithm: Dynamic Programming with Prefix Sums
Techniques: 
- Prefix sums for efficient subarray sum calculation
- HashMap (defaultdict) for counting occurrences
- Iterative construction of valid combinations

Time Complexity: O(5 * n * max_len) where max_len is at most 5, so effectively O(n)
Space Complexity: O(n) for the dictionary storage

This solution groups ticket pieces by their total digit sum and length,
then for each piece, it calculates how many other pieces can be concatenated
to form a lucky ticket.
"""

from collections import defaultdict

MAX_DIGITS = 9


def solve(N, L):

    # Initialize data structures for tracking whole and partial sums
    whole = [defaultdict(int) for _ in range(1 + MAX_DIGITS)]
    part = [defaultdict(int) for _ in range(1 + MAX_DIGITS)]

    for piece in L:
        s = len(piece)
        prefix = [0]

        # Build prefix sum array for current piece
        for c in piece:
            prefix.append(prefix[-1] + ord(c) - ord("0"))

        total_sum = prefix[s]

        # Store the total sum for the full piece
        whole[s][total_sum] += 1

        # Handle all possible splits (for odd lengths, there's a middle element)
        part[s][total_sum] += 1

        # For each valid split point, update the partial sums
        for k in range(1, (s + 1) // 2):
            # Update the counts for the adjusted difference between halves
            part[s - 2 * k][total_sum - 2 * prefix[k]] += 1
            part[s - 2 * k][2 * prefix[s - k] - total_sum] += 1

    result = 0

    # For each possible length, accumulate valid combinations
    for length in range(1, 1 + MAX_DIGITS):
        for sum_value, count in whole[length].items():
            # Multiply the frequency of current sum with matching sums from part
            result += count * part[length][sum_value]

    return result


if __name__ == "__main__":
    N = int(input())
    L = input().split()
    print(solve(N, L))


# https://github.com/VaHiX/CodeForces/